lldb-eval/parser.cc

```diff
@@ -2709,25 +2709,71 @@ ExprResult Parser::BuildTernaryOp(ExprResult cond, ExprResult lhs,
         location, result_type, std::move(cond), std::move(lhs), std::move(rhs));
   }
 
-  // If one operand is a pointer and the other one is arithmeric, convert
-  // arithmetic operand to a pointer.
-  if ((lhs_type.IsPointerType() || lhs_type.IsNullPtrType()) &&
-      (rhs_type.IsScalarOrUnscopedEnum() || rhs_type.IsNullPtrType())) {
+  // If one operand is nullptr and the other one is literal zero, convert
+  // the literal zero to a nullptr type.
+  if (lhs_type.IsNullPtrType() && rhs->is_literal_zero()) {
     rhs = std::make_unique<CStyleCastNode>(
-        rhs->location(), lhs_type, std::move(rhs), CStyleCastKind::kPointer);
+        rhs->location(), lhs_type, std::move(rhs), CStyleCastKind::kNullptr);
 
     return std::make_unique<TernaryOpNode>(location, lhs_type, std::move(cond),
                                            std::move(lhs), std::move(rhs));
   }
-  if ((rhs_type.IsPointerType() || rhs_type.IsNullPtrType()) &&
-      (lhs_type.IsScalarOrUnscopedEnum() || lhs_type.IsNullPtrType())) {
+  if (rhs_type.IsNullPtrType() && lhs->is_literal_zero()) {
     lhs = std::make_unique<CStyleCastNode>(
-        lhs->location(), rhs_type, std::move(lhs), CStyleCastKind::kPointer);
+        lhs->location(), rhs_type, std::move(lhs), CStyleCastKind::kNullptr);
 
     return std::make_unique<TernaryOpNode>(location, rhs_type, std::move(cond),
                                            std::move(lhs), std::move(rhs));
   }
 
+  // At this point there's only one valid scenario left: the resulting type is
+  // a pointer and at least one operand is either an array or a pointer (while
+  // the other operand is an array, a pointer, a nullptr or a literal zero).
+  // Everything else should result with the "incompatible operand types" error.
+
+  // Convert array operands to pointers.
+  if (lhs_type.IsArrayType()) {
+    lhs_type = lhs_type.GetArrayElementType().GetPointerType();
+    lhs = std::make_unique<CStyleCastNode>(
+        lhs->location(), lhs_type, std::move(lhs), CStyleCastKind::kPointer);
+  }
+  if (rhs_type.IsArrayType()) {
+    rhs_type = rhs_type.GetArrayElementType().GetPointerType();
+    rhs = std::make_unique<CStyleCastNode>(
+        rhs->location(), rhs_type, std::move(rhs), CStyleCastKind::kPointer);
+  }
```

Comment:
Actually it seems that according to the C++ standard array-to-pointer conversion happens before everything else. This doesn't change much, but if we do that we can simplify this method even more.

If we apply `array-to-pointer` in the beginning before `if (CompareTypes(...))`, we can handle the case when both operands are the same array there. As a bonus, this will also fix a corner case we currently have -- the result type of `cond ? arr1 : arr2` should be _pointer_, but at the moment we'll have array.

And then later we can do just two more checks:
```
if (lhs_type.IsPointerType() && (rhs_type.IsNullPtrType() || rhs->is_literal_zero()) {
    rhs = std::make_unique<CStyleCastNode>(...)
    return std::make_unique<TernaryOpNode>(...);
}
if ((lhs_type.IsNullPtrType() || lhs->is_literal_zero() && rhs_type.IsPointerType() ) {
    lhs = std::make_unique<CStyleCastNode>(...)
    return std::make_unique<TernaryOpNode>(...);
}
```

I would put them before the case for `nullptr <-> 0`, as it likely a lot more common.